lib/widgets/message_list.dart

```diff
@@ -458,6 +462,67 @@ class ScrollToBottomButton extends StatelessWidget {
   }
 }
 
+class TypingStatusWidget extends StatefulWidget {
+  const TypingStatusWidget({super.key, required this.narrow});
+
+  final Narrow narrow;
+
+  @override
+  State<StatefulWidget> createState() => _TypingStatusWidgetState();
+}
+
+class _TypingStatusWidgetState extends State<TypingStatusWidget> with PerAccountStoreAwareStateMixin<TypingStatusWidget> {
+  TypingStatus? model;
+
+  @override
+  void onNewStore() {
+    model?.removeListener(_modelChanged);
+    model = PerAccountStoreWidget.of(context).typingStatus
+      ..addListener(_modelChanged);
+  }
+
+  @override
+  void dispose() {
+    model?.removeListener(_modelChanged);
+    super.dispose();
+  }
+
+  void _modelChanged() {
+    setState(() {
+      // The actual state lives in [model].
+      // This method was called because that just changed.
+    });
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final store = PerAccountStoreWidget.of(context);
+    final narrow = widget.narrow;
+    if (narrow is! SendableNarrow) return const SizedBox();
+
+    final localizations = ZulipLocalizations.of(context);
+    final typistIds = model!.typistIdsInNarrow(narrow).toList();
+    if (typistIds.isEmpty) return const SizedBox();
+    final String text = switch (typistIds) {
+      [final firstTypist] =>
```

Comment:
https://github.com/zulip/zulip-flutter/pull/790#discussion_r1692365513 (a reply I made yesterday to a question after your revision — echoing here just so it remains visible in this review)

Reply:
Rewritten that part of code. It ended up looking a bit verbose, but it does save the `.toList` call or unnecessary `.map`'s.

Reply:
Great. It may look a little verbose… but I think in actual line count (for this build method) it's the shortest version yet :slightly_smiling_face: Sometimes just writing things very concretely pays off.